### ISSUE TITLE
V's Postgres access is slow compared to Go

### ISSUE THREAD

**Issue by chrisdavies** [vlang/v#14073]
I'm not sure if it's the connection, disconnection, or query, but in my tests, V's Postgres support was about 1/8th the speed of Go's pgx library.

## Details

I ran a little CGI experiment last week.

- For a simple "hello world", V outperformed Go (V's startup time is faster)
  - Go: 4.3K requests per second
  - V: 6.6K requests per second
- When doing a simple DB query, though:
  - Go: 746 request per second
  - V: 85 requests per second

In non CGI-mode, Go was getting many thousands of requests per second, but I didn't test V in that scenario.

Here's the V code:

```v
import pg
import os

fn main() {
	mut cfg := pg.Config{
		host: '127.0.0.1'
		port: 5432
		user: 'xxx'
		password: 'xxx'
		dbname: 'xxx'
	}
	conn := pg.connect(cfg) or {
		println('<h1>500 failed to connect</h1>')
		return
	}
	res := conn.q_string('SELECT id FROM users LIMIT 1') ?
	println('<h1>Hello, $res!</h1>')
	conn.close()
}
```

And here's the Go code:

```go
package main

import (
	"context"
	"fmt"
	"os"

	"github.com/jackc/pgx/v4"
)

func main() {
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, os.Getenv("DB"))
	if err != nil {
		fmt.Println("<h1>500 db conn")
		os.Exit(1)
		return
	}
	defer conn.Close(ctx)
	var id string
	if err := conn.QueryRow(ctx, "SELECT id FROM users LIMIT 1").Scan(&id); err != nil {
		fmt.Println("<h1>500 db query</h1>")
		os.Exit(2)
		return
	}
	fmt.Println("<h1>Hello, " + id + "!</h1>")
}
```

## OS / etc

- Fedora 35
- 16GB RAM
- Dell XPS 15 (from 2020)
- Postgres was running on the same machine

## Pooling?

Incidentally, this brings up one more thing that I didn't see in V, which is connection pooling. In
[pgx](https://github.com/jackc/pgx/tree/master/pgxpool), this is the recommended way to use Postgres
in long-running servers, and it's much faster than opening / closing connections per request.

```go
dbpool, err := pgxpool.Connect(ctx, dbURL)
```

Is there an equivalent in V?

**Comment by kahsa** [comment]
Have you tried `-prod`?

**Comment by chrisdavies** [comment]
No. But I just did, and it was approximately the same:

- hey -c 100 -n 1000 http://127.0.0.1:8000/hello-v
- hey -c 100 -n 1000 http://127.0.0.1:8000/hello-go

```
Requests per second (for 3 independent runs):

V: 92, 90, 91
Go: 885, 883, 902
```

**Comment by medvednikov** [comment]
Thanks for the detailed report @chrisdavies !

Will fix asap.

-prod wouldn't help here, as expected. There's an issue in the pg driver somewhere.

**Comment by FrankLIKE** [comment]
@chrisdavies your  v code is repeat.
mut cfg := pg.Config{
		mut cfg := pg.Config{

**Comment by chrisdavies** [comment]
Yah. That was a copy / paste error when creating the issue. Fixed it.

**Comment by chrisdavies** [comment]
I decided to follow up on this now that 0.3 is out. Here's my update.

Ignoring the connection, and just running the query yields a smaller difference:

10K: queries using an existing (already authenticated) connection

- 440ms in V
- 340ms in Go

The query I ran was this, where id is a UUID column:

```v
// V
res := conn.q_string('SELECT id FROM users LIMIT 1') ?
```

```go
// Go
// Using "github.com/jackc/pgx/v4" without a connection pool so the comparison is apples-to-apples
err := conn.QueryRow(ctx, "SELECT id FROM users LIMIT 1").Scan(&id)
```

Note, there wasn't a significant difference between querying a UUID column vs a VARCHAR.

This got me wondering if the cost of the connection was causing my previous poor result, but no:

1K: connect, query, close connection

- 10.7s in V
- 14.4s in V

I'm not sure why my results are different this time. There are some differences from the last run I made:

- Go 1.18 vs 1.17 (or maybe 1.16)
- Postgres 14 instead of 13
- V 0.3 instead of 0.2

I'd say that, if used with a connection pool, V's current Postgres performance is good enough for building services and close enough to Go that we might close this issue.

I think I'm going to spin up a side project in V.

My test were run on Fedora 36, Postgres 14 (running locally, no TLS or anything),  Intel Core i7-9750H.
